cryptoki/tests/basic.rs

```diff
@@ -723,3 +723,61 @@ fn aes_key_attributes_test() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+#[serial]
+fn ro_rw_session_test() -> Result<()> {
+    let public_exponent: Vec<u8> = vec![0x01, 0x00, 0x01];
+    let modulus = vec![0xFF; 1024];
+
+    let template = vec![
+        Attribute::Token(true),
+        Attribute::Private(false),
+        Attribute::PublicExponent(public_exponent),
+        Attribute::Modulus(modulus),
+        Attribute::Class(ObjectClass::PUBLIC_KEY),
+        Attribute::KeyType(KeyType::RSA),
+        Attribute::Verify(true),
+    ];
+
+    let (pkcs11, slot) = init_pins();
+
+    // Try out Read-Only session
+    {
+        // open a session
+        let ro_session = pkcs11.open_ro_session(slot)?;
+
+        // log in the session
+        ro_session.login(UserType::User, Some(USER_PIN))?;
+
+        // generate a key pair
+        // This should NOT work using the Read/Write session
+        let e = ro_session.create_object(&template).unwrap_err();
+
+        if let Error::Pkcs11(RvError::SessionReadOnly) = e {
```

Comment:
Minor: why not use `assert_eq!` ?

Reply:
Because I can't derive `PartialEq` on `Error`, as it contains some types that don't implement it 😞 So I'd have to use an `if let` anyway.